### DISCUSSION TITLE
It would be useful if ChatCompletionMessage could be used everywhere a chat message is needed.

### DISCUSSION THREAD

**Discussion by zby** [openai/openai-python#1358] [Ideas]
Currently you cannot use `ChatCompletionMessage` as a tool result message:
```
message_dict = {
     "tool_call_id": 'tool_call.id',
    "role": "tool",
    "name": 'function_name',
    "content": 'function_response',
}

message = ChatCompletionMessage(**message_dict)
```
OUTPUT:
```
Traceback (most recent call last):
  File "/home/zby/gpt/answerbot/tmp/ChatCompletionMessage_example.py", line 10, in <module>
    message = ChatCompletionMessage(**message_dict)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/zby/gpt/answerbot/.venv/lib/python3.11/site-packages/pydantic/main.py", line 175, in __init__
    self.__pydantic_validator__.validate_python(data, self_instance=self)
pydantic_core._pydantic_core.ValidationError: 1 validation error for ChatCompletionMessage
role
  Input should be 'assistant' [type=literal_error, input_value='tool', input_type=str]
    For further information visit https://errors.pydantic.dev/2.7/v/literal_error
```

This is for saving the tool result in chat history - like in the python example at https://platform.openai.com/docs/guides/function-calling

Using objects in the chat history seems more robust than using raw dictionaries.

**Comment by rattrayalex** [comment]
Hey @zby , just wanted to say thanks for raising this. We're mulling it over. It'd be helpful if you could share more of the code you're trying to write.

**Reply by zby** [reply]
Just to be more explicit. The example code from the docs is:
```
            messages.append(
                {
                    "tool_call_id": tool_call.id,
                    "role": "tool",
                    "name": function_name,
                    "content": function_response,
                }
            )  # extend conversation with function response
        second_response = client.chat.completions.create(
            model="gpt-3.5-turbo-0125",
            messages=messages,
        ) 
```
For other messages I often appended a `ChatCompletionMessage` object to the messages list instead of the dictionary that represents the message, when I have such an object from a previous LLM response. This worked and seemed more robust than passing dictionaries - but apparently a tool result cannot be a ChatCompletionMessage.

In other words - the messages argument to client.chat.completions.create could be of type list[ChatCompletionMessage] instead of list[dictionary].

I don't have any code left from the experiment where I tried appending the tool response as an object to the messages list, I am not even sure if I ever saved it to the repo.
